### PR TITLE
Ability to export logs to lang

### PR DESCRIPTION
**Pull request by Sushisource**
## What was changed
Adds a facility to export logs to lang so that they can show up with the user's configured loggers. Loses some context from tracing but overall still useful.

## Why?
Easy-to-enable debuggability for users

Relates to https://github.com/temporalio/sdk-core/issues/85

### CODE REVIEW

File: src/telemetry/mod.rs [hunk 1]
```diff
@@ -28,21 +31,31 @@ pub struct TelemetryOptions {
...
     /// variable.
     pub tracing_filter: String,
+    /// Core can forward logs to lang for them to be rendered by the user's logging facility.
+    /// The logs are somewhat contextually lacking, but still useful in a local test situation when
+    /// running only one workflow at a time. This sets the level at which they are filtered.
+    /// TRACE level will export span start/stop info as well.
+    ///
+    /// Default is INFO. If set to `Off`, the mechanism is disabled entirely (saves on perf).
+    /// If set to anything besides `Off`, any console output directly from core is disabled.
+    pub log_forwarding_level: LevelFilter,
```
Comment (bergundy) [inline src/telemetry/mod.rs:41]: So basically to turn off logging completely we need to set `log_forwarding_level` to `Off` and `tracing_filter` to `"OFF"`?

File: src/telemetry/mod.rs [hunk 2]
```diff
@@ -28,21 +31,31 @@ pub struct TelemetryOptions {
     /// telemetry or console output. May be overridden by the `TEMPORAL_TRACING_FILTER` env
```
Comment (bergundy) [inline src/telemetry/mod.rs:31]:
```suggestion
    /// telemetry, forwarding to lang, or console output. May be overridden by the `TEMPORAL_TRACING_FILTER` env
```

File: src/telemetry/mod.rs (hunk 1 above)
Comment (bergundy) [reply src/telemetry/mod.rs:41]: I think we might want to have 2 filters instead and 1 boolean:
1. filter for otel collector
2. filter for output / forwarding
3. boolean toggle for forwarding / output directly to console

That'd be the most flexible, is it even possible?

File: src/log_export.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+pub(crate) struct CoreExportLogger {
+    records: Mutex<VecDeque<CoreLog>>,
+    level_filter: LevelFilter,
+}
+
+impl CoreExportLogger {
+    pub(crate) fn new(level: LevelFilter) -> Self {
+        Self {
+            records: Mutex::new(Default::default()),
```
Comment (bergundy) [inline src/log_export.rs:38]: We mentioned limiting the capacity of the buffer, I think it'd be better to do that.

**Review by bergundy — APPROVED** [review]
This is almost good to merge, left a few comments, PTAL before merging.

File: src/telemetry/mod.rs (hunk 1 above)
Comment (Sushisource) [reply src/telemetry/mod.rs:41]: I don't think there's any reason to forward logs *and* have them go direct to console from core, you'll just get duplicated output. So with that in mind the two filters pretty much do what you need.

Comment (bergundy) [reply src/telemetry/mod.rs:41]: But there's only 1 filter.
How do I configure ">=error to console and >=info to the collector"?

File: src/log_export.rs [hunk 4]
```diff
@@ -28,24 +28,35 @@ impl CoreLog {
...
 pub(crate) struct CoreExportLogger {
-    records: Mutex<VecDeque<CoreLog>>,
+    logs_in: Mutex<Producer<CoreLog>>,
+    logs_out: Mutex<Consumer<CoreLog>>,
     level_filter: LevelFilter,
 }
 
 impl CoreExportLogger {
     pub(crate) fn new(level: LevelFilter) -> Self {
+        let (lin, lout) = RingBuffer::new(2048).split();
```
Comment (bergundy) [inline src/log_export.rs:38]: This should be configurable

File: src/log_export.rs [hunk 5]
```diff
@@ -64,10 +75,11 @@ impl Log for CoreExportLogger {
             timestamp: SystemTime::now(),
             level: record.level(),
         };
-        self.records
+        let _ = self
+            .logs_in
             .lock()
             .expect("Logging mutex must be acquired")
-            .push_back(clog)
+            .push(clog);
```
Comment (bergundy) [inline src/log_export.rs:82]: There should be a way to know if we are losing messages due to slow consumer.

**Pull request merged**